be/src/runtime/thread_context.h

```diff
@@ -299,33 +300,87 @@ class StopCheckThreadMemTrackerLimit {
     tracker->transfer_to(                               \
             size, doris::thread_context()->_thread_mem_tracker_mgr->limiter_mem_tracker_raw())
 
+// Consider catching other memory errors, such as memset failure, etc.
+#define RETURN_IF_CATCH_BAD_ALLOC(stmt)                                                          \
+    do {                                                                                         \
+        doris::thread_context()->_thread_mem_tracker_mgr->clear_exceed_mem_limit_msg();          \
+        if (doris::thread_context()->enable_thread_cache_bad_alloc) {                            \
+            try {                                                                                \
+                { stmt; }                                                                        \
+            } catch (std::bad_alloc const&) {                                                    \
+                return Status::MemoryLimitExceeded(                                              \
+                        fmt::format("PreCatch Bad Alloc: {}",                                    \
+                                    doris::thread_context()                                      \
+                                            ->_thread_mem_tracker_mgr->exceed_mem_limit_msg())); \
+            }                                                                                    \
+        } else {                                                                                 \
+            try {                                                                                \
+                doris::thread_context()->enable_thread_cache_bad_alloc = true;                   \
```

Comment:
using Defer defer  here.  because if user add a return stmt in params the context is not set back.